### PR TITLE
DDF-3798: Users should be prevented from running DDF under specific when security manager is enabled

### PR DESCRIPTION
**Pull request by figliold**
#### What does this PR do?
Prevents users from running DDF under specific directories when security manager is enabled

#### Who is reviewing it? 
@tbatie 
@emanns95 
@paouelle 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Choose 2 committers to review/merge the PR. 
@clockard
@lessarderic 
#### How should this be tested? (List steps with links to updated documentation)
1) Unzip DDF in your user.home directory and verify that it shutsdown automatically with an error message stating that DDF cannot be installed in user.home.
2) Unzip DDF in another directory outside of user.home, and verify that it starts up and runs with no problems.
#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-3798
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/secure-boot/pom.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                    </instructions>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
```
Comment (clockard) [inline platform/security/secure-boot/pom.xml]: Since this is over the 75% should we omit this plugin overrided for this bundle?

File: distribution/docs/src/main/resources/content/_installing/installing-intro.adoc [hunk 2]
```diff
@@ -23,6 +23,11 @@ Adding other components does not have any security/hardening implications.
 
 === Installation Prerequisites
 
+[WARNING]
```
Comment (paouelle) [inline distribution/docs/src/main/resources/content/_installing/installing-intro.adoc]: ❓ ❗️ Will that be inherited by all upstream projects? If so than it would be a problem for those that do not enable the security manager.

File: distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc [hunk 3]
```diff
@@ -22,6 +22,11 @@ These steps will demonstrate:
 
 These are the basic requirements to set up the environment to run a ${branding}.
 
+[WARNING]
```
Comment (paouelle) [inline distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc]: ❓ ❗️ Will that be inherited by all upstream projects? If so than it would be a problem for those that do not enable the security manager.

File: platform/security/secure-boot/pom.xml (hunk 1 above)
Comment (paouelle) [reply platform/security/secure-boot/pom.xml]: Keeping it like this will ensure that anybody adding new code won't simply skip test cases because the current ones already account for the 75% minimum.

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    try {
+      ddfHome = getDdfHome();
+      userHome = getUserHome();
+    } catch (IOException e) {
+      String message =
+          "Unable to determine if your installation ["
+              + ddfHome
+              + "] is located in an insecure directory. See log for details. Shutting down...";
+      System.err.println("\n" + message + "\n");
+      LOGGER.error(message, e);
```
Comment (paouelle) [inline platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: ❓ Do we want to log exceptions and their stack trace at the ERROR level? Should the stack trace be done at DEBUG instead?

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+  private static final String NOW = "0";
+
+  private final SystemService systemService;
+
+  private Path ddfHome;
+
+  private Path userHome;
+
+  private boolean shuttingDown = false;
```
Comment (paouelle) [inline platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: ❗️ Should be volatile

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  boolean securityManagerEnabled() {
+    return System.getSecurityManager() != null;
+  }
+
+  @VisibleForTesting
+  void systemExit(Exception e) {
+    String message =
+        "ERROR: Exception encountered while shutting system down via SystemService. Terminating JVM...";
+    System.err.println("\n" + message + "\n");
+    LOGGER.error(message, e);
```
Comment (paouelle) [inline platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: Same comment as before about logging the stack trace at the ERROR level

File: platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        then:
+        1 * systemService.halt("0") >> { throw new Exception("unable to halt")}
+        exitedDueToSystemServiceFailureToExit == true
+    }
+
+    def 'System exits normally when attempting to read invalid ddf.home'() {
+        setup:
+        def exitedDueToSystemServiceFailureToExit = false
+        def userHomeSysProp = userHome.getRoot().toString()
+        def ddfHomeSysProp = "/" + System.currentTimeMillis() + "/" + UUID.randomUUID() + "/ddf"
```
Comment (paouelle) [inline platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy]: ❓will that work on Windows?

File: platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    /**
+     * ie. /opt/ddfHomeSymLink -> /users/tomPenny/ddf
+     */
+    def 'System exits normally when security manager enabled and symbolic link outside of user home dir pointing to ddf installation inside of user home dir'() {
+        setup:
+        def exitedDueToSystemServiceFailureToExit = false
+        def optDir = temporaryFolder.newFolder("opt").toPath()
+        def tomPennyHomeDir = temporaryFolder.newFolder("tomPenny")
+        def ddfHomeDir = Files.createDirectory(Paths.get(tomPennyHomeDir.toString(), "ddf"))
+        def ddfHomeSymLink = Files.createSymbolicLink(Paths.get(optDir.toString(), "ddfHomeSymLink"), ddfHomeDir)
```
Comment (paouelle) [inline platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy]: ❓ Does it work on WIndows? symlinks testing on Windows are always a big pain.

File: distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc (hunk 3 above)
Comment (tbatie) [reply distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc]: I think we should consider security manager a best practice and less of a feature that can be turned on and off. I would argue if upstream projects are not using the security manager, that it's on them to fix it, not for DDF to support extensibility points in the scenario that they turn it off.

**Comment by figliold** [comment]
build now

File: distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc (hunk 3 above)
Comment (figliold) [reply distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc]: I also agree that using the security manager should be considered the norm not the exception.  I will leave the documentation as-is.

File: distribution/docs/src/main/resources/content/_installing/installing-intro.adoc (hunk 2 above)
Comment (figliold) [reply distribution/docs/src/main/resources/content/_installing/installing-intro.adoc]: I think that using the security manager should be considered the norm not the exception.  I will leave the documentation as-is.

File: platform/security/secure-boot/pom.xml (hunk 1 above)
Comment (figliold) [reply platform/security/secure-boot/pom.xml]: I leave it in.

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java (hunk 5 above)
Comment (figliold) [reply platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: done

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java (hunk 4 above)
Comment (figliold) [reply platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: fixed

File: platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java (hunk 6 above)
Comment (figliold) [reply platform/security/secure-boot/src/main/java/org/codice/ddf/security/boot/SecureBoot.java]: fixed

File: platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy (hunk 7 above)
Comment (figliold) [reply platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy]: I have tested these on Windows and they work.

File: platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy (hunk 8 above)
Comment (figliold) [reply platform/security/secure-boot/src/test/groovy/org/codice/ddf/security/boot/SecureBootSpec.groovy]: I have tested these on Windows and they work.

**Comment by figliold** [comment]
flaky itests; build now

File: distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc (hunk 3 above)
Comment (paouelle) [reply distribution/docs/src/main/resources/content/_quickstart/quickstart-installing.adoc]: For now, that is good enough for me. I just wanted to make sure we were aware of the implications. We can revisit if we ever see DDF use in non-defense type of applications where such tight security is not required.

**Comment by tbatie** [comment]
Hero'd successfully. Verified the distribution started up as expected outside the home dir. When in the home dir, the distribution shut down with the expected error message. Also verified the distribution runs properly under the home dir when the security manager is disabled.

**Comment by figliold** [comment]
build now

**Comment by figliold** [comment]
build now
